### PR TITLE
feat(ls): ignore ci ignored

### PR DESCRIPTION
**Pull request by ajbt200128**
Previously the language server would show findings that have been ignored in CI. This PR makes it so if they're ignored in ci they don't show in the extensions.

This PR also fixes a bug where we weren't deduplicating findings for the LS.

Finally this PR also ports match based ID to OSemgrep! See the comments about it though, it's in an experimental state.

## Test Plan

There's a unit test :)

PR checklist:

- [X] Purpose of the code is [evident to future readers](https://semgrep.dev/docs/contributing/contributing-code/#explaining-code)
- [X] Tests included or PR comment includes a reproducible test plan
- [X] Documentation is up-to-date
- [X] A changelog entry was [added to changelog.d](https://semgrep.dev/docs/contributing/contributing-code/#adding-a-changelog-entry) for any user-facing change
- [X] Change has no security implications (otherwise, ping security team)

If you're unsure about any of this, please see:

- [Contribution guidelines](https://semgrep.dev/docs/contributing/contributing-code)!
- [One of the more specific guides located here](https://semgrep.dev/docs/contributing/contributing/)

### CODE REVIEW

**Comment by ajbt200128** [comment]
It looks like for some reason there are some rules who's messages aren't being pulled in. Will investigate tomorrow, but a bandaid fix has been pushed

File: changelog.d/lang-server.added [hunk 1]
```diff
@@ -0,0 +1 @@
+Semgrep LS will now not show findings that have been ignored in CI
```
Comment (aryx) [inline changelog.d/lang-server.added]: I'm not sure user are familiar with what LS mean. Maybe say Semgrep language server, or
Semgrep LSP (I know we should use LS, because LSP is the protocol, but I think most people
are familiar with what LSP mean but not LS).

Comment (aryx) [reply changelog.d/lang-server.added]: What means CI here? You mean ignored in the Semgrep App?

File: src/osemgrep/cli_scan/Output.ml [hunk 2]
```diff
@@ -151,20 +151,11 @@ let dispatch_output_format (output_format : Output_format.t)
...
- * TODO: take a more precise conf than Scan_CLI.conf at some point
- *)
-let output_result (conf : Scan_CLI.conf) (profiler : Profiler.t)
-    (res : Core_runner.result) : Out.cli_output =
-  (* In theory, we should build the JSON CLI output only for the
-   * Json conf.output_format, but cli_output contains lots of data-structures
-   * that are useful for the other formats (e.g., Vim, Emacs), so we build
-   * it here.
-   *)
+let preprocess_result (conf : Scan_CLI.conf) (res : Core_runner.result) :
```
Comment (aryx) [inline src/osemgrep/cli_scan/Output.ml:160]: maybe a small comment before introducing what this function does?

File: src/osemgrep/cli_scan/Output.ml [hunk 3]
```diff
@@ -175,9 +166,29 @@ let output_result (conf : Scan_CLI.conf) (profiler : Profiler.t)
...
+ * TODO: take a more precise conf than Scan_CLI.conf at some point
+ *)
+let output_result (conf : Scan_CLI.conf) (profiler : Profiler.t)
+    (res : Core_runner.result) : Out.cli_output =
+  (* In theory, we should build the JSON CLI output only for the
+   * Json conf.output_format, but cli_output contains lots of data-structures
+   * that are useful for the other formats (e.g., Vim, Emacs), so we build
+   * it here.
+   *)
+  let cli_output = preprocess_result conf res in
```
Comment (aryx) [inline src/osemgrep/cli_scan/Output.ml]: this is actually not a cli_output, it's a function returning a cli_output.

File: src/osemgrep/cli_scan/Output.ml [hunk 4]
```diff
@@ -175,9 +166,29 @@ let output_result (conf : Scan_CLI.conf) (profiler : Profiler.t)
...
+  (* TOPORT? output.output() *)
   let cli_output = Profiler.record profiler ~name:"ignores_times" cli_output in
```
Comment (aryx) [inline src/osemgrep/cli_scan/Output.ml:192]: why not removing this unit argument in preprocess_result and instead just do here
let cli_output = Profile.record profile ... (fun () ->
    preprocess_result conf res)
in

File: src/osemgrep/cli_scan/Output.mli [hunk 5]
```diff
@@ -1,5 +1,13 @@
 module Out = Semgrep_output_v1_j
 
+val preprocess_result :
+  Scan_CLI.conf -> Core_runner.result -> unit -> Out.cli_output
+(** [preprocess_result conf result] preprocesses the result of a scan
+  * according to the configuration [conf]. This handles
+  * nosemgrep, interolating messages, and more. It returns a function that
```
Comment (aryx) [inline src/osemgrep/cli_scan/Output.mli:6]: I would remove the unit parameter and related comment.
The need to profile function should not require to change their signature. It just adds an extra
complexity that is not needed.

File: src/osemgrep/language_server/Processed_run.ml [hunk 6]
```diff
@@ -44,20 +44,19 @@ let filter_clean_lines matches =
...
-  let matches = result.core.results in
-  let hrules = result.hrules in
+let of_matches ?(skipped_fingerprints = []) ?(only_git_dirty = true)
+    (result : Core_runner.result) =
+  let result = (Output.preprocess_result scan_conf result) () in
   (* Match the rules with the matches so we can get fixes/rule-ids/messages *)
   let matches =
-    matches
-    |> Common.map (Cli_json_output.cli_match_of_core_match hrules)
-    |> Common.exclude (fun m ->
```
Comment (aryx) [inline src/osemgrep/language_server/Processed_run.ml:58]: why deleting this part? Shouldn't we just add an extra filter but keep the original code?

File: src/osemgrep/language_server/Processed_run.mli [hunk 7]
```diff
@@ -1,4 +1,5 @@
 val of_matches :
```
Comment (aryx) [inline src/osemgrep/language_server/Processed_run.mli:1]: small comment maybe?

File: src/osemgrep/language_server/Session.mli [hunk 8]
```diff
@@ -1,8 +1,12 @@
 open Lsp
 open Types
 
-type rule_cache = { mutable rules : Rule.t list; lock : Lwt_mutex.t }
-(** Cache of active rules. Protected by mutex as [cache_rules] below can be called asynchronously,
+type session_cache = {
+  mutable rules : Rule.t list;
```
Comment (aryx) [inline src/osemgrep/language_server/Session.mli:5]: maybe some comment before each field explaining the field instead of this global comment at the end of the record?
That would also give the opportunity to explain what "active" mean, and
where the skipped_fingerprints come from.

File: src/osemgrep/language_server/Unit_LS.ml [hunk 9]
```diff
@@ -1,15 +1,16 @@
-open Lsp
-open Types
 open Testutil
```
Comment (aryx) [inline src/osemgrep/language_server/Unit_LS.ml:1]: thx for removing some open!

File: src/osemgrep/language_server/Unit_LS.ml [hunk 10]
```diff
@@ -91,6 +92,17 @@ let add_file ?(git = false) ?(dirty = false)
   if (not dirty) && git then Git_wrapper.commit workspace "test";
   file
 
+let with_mock_envvars f () =
+  let old_settings = !Semgrep_envvars.v in
+  let new_settings = { old_settings with app_token = Some "123456789" } in
+  Semgrep_envvars.v := new_settings;
```
Comment (aryx) [inline src/osemgrep/language_server/Unit_LS.ml]: You can use Common.save_excursion (which is also exn-friendly)

File: src/osemgrep/language_server/Unit_LS.ml [hunk 11]
```diff
@@ -222,29 +234,46 @@ let processed_run () =
...
+let ci_tests () =
   let with_ci_client =
     let make_fn (req : Cohttp.Request.t) body =
       ignore body;
+      let uri = req |> Cohttp.Request.uri |> Uri.path in
       Http_mock_client.check_method `GET req.meth;
+      let body_file =
+        match uri with
+        | "/api/agent/deployments/scans/config" ->
+            "./tests/ls/ci/rule_conf_resp.json"
```
Comment (aryx) [inline src/osemgrep/language_server/Unit_LS.ml]: Again I think it would better to inline their content here, like let body reply weither with a File of Fpath.t or a Content of string

File: src/osemgrep/networking/Semgrep_App.ml [hunk 12]
```diff
@@ -9,69 +9,65 @@ open Common
  *)
 
 (*****************************************************************************)
-(* Entry points *)
+(* Types *)
 (*****************************************************************************)
+(* TODO? specify this with atd and have both app + osemgrep use it *)
```
Comment (aryx) [inline src/osemgrep/networking/Semgrep_App.ml:14]: YES! We can do that in a separate PR if you prefer, but we must remember to do it ASAP!

File: src/osemgrep/reporting/Cli_json_output.ml [hunk 13]
```diff
@@ -292,6 +292,61 @@ let cli_error_of_core_error (x : Out.core_error) : Out.cli_error =
 (*****************************************************************************)
 (* LATER: we should get rid of those intermediate Out.core_xxx *)
 
+(* This is a cursed function that calculates everything but the index part
+ * of the match_based_id. It's cursed because we need hashes to be exactly
+ * the same, but the algorithm used on the python side to generate
+ * the final string thats hashed has some python specific quirks.
```
Comment (aryx) [inline src/osemgrep/reporting/Cli_json_output.ml:298]: Ah! I told you :)

Comment (aryx) [reply src/osemgrep/reporting/Cli_json_output.ml:298]: Still I don't understand the use of "cursed" here.

File: src/osemgrep/reporting/Cli_json_output.ml [hunk 14]
```diff
@@ -292,6 +292,61 @@ let cli_error_of_core_error (x : Out.core_error) : Out.cli_error =
...
+ * (see https://github.com/returntocorp/semgrep/blob/2d34ce584d16c4e954349690a5f12fae877a94d6/cli/src/semgrep/rule.py#L289-L334)
+ * 1. Sort all top level keys (i.e pattern, patterns etc.) alphabetically
+ * 2. For each key: DFS the tree and find all pattern values (i.e. the rhs of pattern: <THING>)
+ * 3. Sort all pattern values alphabetically and concatenate them with a space
+ * 4. Concatenate all the sorted pattern values with a space
+ * 5. Hash the tuple `(sorted_pattern_values, path, rule_id)` w/ blake2b
+ * 6. Append the index of the match in the list of matches for the rule (see [index_match_based_ids])
+ *
+ * Austin: I wrote the initial match based ID, and this one below is a port of it.
+ * Looking back it seems like I ended up writing a roundabout version of the below algorithm
```
Comment (aryx) [inline src/osemgrep/reporting/Cli_json_output.ml:311]: what roundabout means here?

File: src/osemgrep/reporting/Cli_json_output.ml [hunk 15]
```diff
@@ -292,6 +292,61 @@ let cli_error_of_core_error (x : Out.core_error) : Out.cli_error =
...
+ * 3. Sort all of step 2 results alphabetically and concatenate them with a space
+ * 4. Same as step 5 and 6
+ *
+ * Assumptions:
+ * Somewhere it seems that all keys are already sorted alphabetically when the rule is parsed.
+ * I have not seen this code. I simply have faith that it is true and this will not change.
+ *
+ * I'm also hoping that [interpolate_metavariables] works similar to what we do on the python side
+ *
+ * I have not tested this code beyond checking a bunch of examples manually.
```
Comment (aryx) [inline src/osemgrep/reporting/Cli_json_output.ml:324]: we currently mask the fingerprint to 0x42 in conftest.py I think, so it's easy to test if it works, remove the masking code and we can see whether we have many regressions when running make osemgrep-e2e.

File: src/osemgrep/reporting/Cli_json_output.ml [hunk 16]
```diff
@@ -292,6 +292,61 @@ let cli_error_of_core_error (x : Out.core_error) : Out.cli_error =
...
+ *
+ * There's some weird thing we do w/ join mode. I am hoping that this doesn't matter irl
```
Comment (aryx) [inline src/osemgrep/reporting/Cli_json_output.ml:326]: yep. We should not care too much about join mode; we will soon delete it as Emma has a replacement for it (steps mode).

File: src/osemgrep/reporting/Cli_json_output.ml [hunk 17]
```diff
@@ -292,6 +292,61 @@ let cli_error_of_core_error (x : Out.core_error) : Out.cli_error =
...
+ *)
+let match_based_id_partial rule rule_id metavars path =
```
Comment (aryx) [inline src/osemgrep/reporting/Cli_json_output.ml:328]: noice!! Excting!!

File: src/osemgrep/reporting/Cli_json_output.ml [hunk 18]
```diff
@@ -390,6 +443,40 @@ let dedup_and_sort (xs : Out.cli_match list) : Out.cli_match list =
            Hashtbl.replace seen key true;
            true)
 
+(* This is the same algorithm for indexing as in cli. We shouldn't need to update this *)
```
Comment (aryx) [inline src/osemgrep/reporting/Cli_json_output.ml]: what is cli in this context? Do you mean pysemgrep?

Comment (aryx) [reply src/osemgrep/reporting/Cli_json_output.ml]: Can you add some comments explaning why we need this function? And what it does?

File: src/osemgrep/reporting/Cli_json_output.mli [hunk 19]
```diff
@@ -14,5 +14,12 @@ val exit_code_of_error_type : Semgrep_output_v1_t.core_error_kind -> Exit_code.t
 val cli_match_of_core_match :
   Rule.hrules -> Semgrep_output_v1_t.core_match -> Semgrep_output_v1_t.cli_match
 
+val index_match_based_ids :
+  Semgrep_output_v1_t.cli_match list -> Semgrep_output_v1_t.cli_match list
+(** [index_match_based_ids matches] will append an index to the match based id
```
Comment (aryx) [inline src/osemgrep/reporting/Cli_json_output.mli:19]: but the match based id is a string right? You mean an index is appended to the string?
Isn't the match_based_id supposed to be a match identifier ... why we need this extra index?

**Review by aryx — APPROVED** [review]

File: changelog.d/lang-server.added (hunk 1 above)
Comment (ajbt200128) [reply changelog.d/lang-server.added]: Yup, meant the app, will update this accordingly

File: src/osemgrep/cli_scan/Output.ml (hunk 4 above)
Comment (ajbt200128) [reply src/osemgrep/cli_scan/Output.ml:192]: Duh, that makes a lot of sense, will change it

File: src/osemgrep/cli_scan/Output.mli (hunk 5 above)
Comment (ajbt200128) [reply src/osemgrep/cli_scan/Output.mli:6]: yea I wasn't thinking, will fix it

File: src/osemgrep/language_server/Processed_run.ml (hunk 6 above)
Comment (ajbt200128) [reply src/osemgrep/language_server/Processed_run.ml:58]: I refactored the preprocess_result code out from output so we could use it here. This way we're sharing code between the scan command and LS, so we apply all the same formating/filtering/deduping etc.

File: src/osemgrep/language_server/Processed_run.mli (hunk 7 above)
Comment (ajbt200128) [reply src/osemgrep/language_server/Processed_run.mli:1]: ? there is one right below it

File: src/osemgrep/language_server/Session.mli (hunk 8 above)
Comment (ajbt200128) [reply src/osemgrep/language_server/Session.mli:5]: will do

File: src/osemgrep/language_server/Unit_LS.ml (hunk 10 above)
Comment (ajbt200128) [reply src/osemgrep/language_server/Unit_LS.ml]: Forgot about this again, will change

File: src/osemgrep/language_server/Unit_LS.ml (hunk 11 above)
Comment (ajbt200128) [reply src/osemgrep/language_server/Unit_LS.ml]: Fair, will change

File: src/osemgrep/networking/Semgrep_App.ml (hunk 12 above)
Comment (ajbt200128) [reply src/osemgrep/networking/Semgrep_App.ml:14]: Def a separate PR, but will add it to my list of tasks

File: src/osemgrep/reporting/Cli_json_output.ml (hunk 13 above)
Comment (ajbt200128) [reply src/osemgrep/reporting/Cli_json_output.ml:298]: Yea, I forgot about the tuple stuff

Comment (ajbt200128) [reply src/osemgrep/reporting/Cli_json_output.ml:298]: This function is cursed because it shouldn't exist, i.e. in a perfect world, we wouldn't need to emulate python things in ocaml.

File: src/osemgrep/reporting/Cli_json_output.ml (hunk 14 above)
Comment (ajbt200128) [reply src/osemgrep/reporting/Cli_json_output.ml:311]: synonym for overly complicated/indirect

File: src/osemgrep/reporting/Cli_json_output.ml (hunk 15 above)
Comment (ajbt200128) [reply src/osemgrep/reporting/Cli_json_output.ml:324]: I'll check that out

File: src/osemgrep/reporting/Cli_json_output.ml (hunk 16 above)
Comment (ajbt200128) [reply src/osemgrep/reporting/Cli_json_output.ml:326]: Cool that's what I figured

File: src/osemgrep/reporting/Cli_json_output.ml (hunk 18 above)
Comment (ajbt200128) [reply src/osemgrep/reporting/Cli_json_output.ml]: Yea, meant pysemgrep will change it and add more of an explanation

File: src/osemgrep/reporting/Cli_json_output.mli (hunk 19 above)
Comment (ajbt200128) [reply src/osemgrep/reporting/Cli_json_output.mli:19]: I'm adding an explanation, but the tldr is we do some logic on the app side for deduplication based on the index of the id.

File: src/osemgrep/reporting/Cli_json_output.ml (hunk 15 above)
Comment (ajbt200128) [reply src/osemgrep/reporting/Cli_json_output.ml:324]: Looks like we could set something up, but would rather do that in a different PR
